File: problems-extend-adapt/3901-4000/3997_counting-the-subtree-champions/solution.py
class Solution:
    def countSubtreeChampions(self, root):
        def go(node):
            if node is None:
                return (-1, 0)
            a, x = go(node.left)
            b, y = go(node.right)
            m = max(node.val, a, b)
            return (m, x + y + (node.val == m))

        return go(root)[1]

File: problems-extend-adapt/3901-4000/3997_counting-the-subtree-champions/solution.rs
impl Solution {
    fn dom(n: &Option<Box<TreeNode>>) -> (i32, i32) {
        if let Some(v) = n {
            let (a, x) = Self::dom(&v.left);
            let (b, y) = Self::dom(&v.right);
            let m = v.val.max(a).max(b);
            (m, x + y + (v.val == m) as i32)
        } else {
            (-1, 0)
        }
    }
    pub fn count_subtree_champions(root: Option<Box<TreeNode>>) -> i32 {
        Self::dom(&root).1
    }
}
